### PR TITLE
feat: add future_deque package with FutureDeque and LocalFutureDeque

### PR DESCRIPTION
**Pull request by sandersaares**
## Summary

Add a new workspace package `future_deque` providing pool-backed deque collections for managing groups of futures with deterministic front-to-back polling order.

## Key types

- **`FutureDeque<T>`** — thread-mobile (`Send`) variant for cross-thread use
- **`LocalFutureDeque<T>`** — single-threaded (`!Send`) variant that accepts any future

Both types support `push_front`/`push_back`, `pop_front`/`pop_back` with strict deque semantics (only the actual front or back can be popped, and only if completed).

Both implement `Future<Output = ()>` (returns `Ready(())` when all contained futures have completed), and optionally `Stream` (behind the `futures-stream` feature).

## Design

- Futures stored in thread-local object pools (`BlindPool`/`LocalBlindPool`) with type erasure via `ErasedFuture` trait
- Per-slot waker metadata stored in a thread-local `PinnedPool` with stable pinned addresses used as `RawWaker` data pointers — zero heap allocations after pool warm-up
- Shared parent waker (`Arc<Mutex<Waker>>`) per deque instance, initialized to `Waker::noop()` and updated once per `poll()`
- Atomic activation flag with first-setter deduplication to avoid redundant parent wakes
- Manual reference counting for waker metadata with safe cross-thread cleanup
- Shared core (`FutureDequeCore<T, H>`) with thin public wrappers enforcing `Send` bounds

## Contents

- 51 unit tests (ordering, edge cases, drop, multithreaded with real `events_once::Event` signaling)
- Static trait assertions (`Send`/`Sync`/`Unpin`/`Future`)
- 6 Criterion benchmarks (2 variants x 3 workloads)
- 2 standalone examples + README

### CODE REVIEW

File: packages/futurism/src/future_deque.rs [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+// and `push_front` both require `F: Future + Send + 'static`, guaranteeing that every
+// value behind the trait object is in fact `Send`. This is the intended usage pattern of
+// `BlindPooledMut` — it deliberately does not require `T: Send` so that trait object
+// casts do not need to carry marker bounds. All other fields (`BlindPool`, waker metadata
+// behind `Arc<Mutex<…>>`) are `Send + Sync`.
+unsafe impl<T: Send> Send for FutureDeque<T> {}
+
+#[cfg(test)]
+#[cfg_attr(coverage_nightly, coverage(off))]
+mod tests {
```
Comment (sandersaares) [inline packages/futurism/src/future_deque.rs]: It feels like we could use more tests here, as there appears to be minimal unit test covergae of this type.

File: packages/futurism/src/deque_future.rs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+};
+
+use infinity_pool::define_pooled_dyn_cast;
+
+/// Type erasure trait for futures stored in [`FutureDeque`][crate::FutureDeque].
+///
+/// We cannot use `Future<Output = T>` directly with `define_pooled_dyn_cast!` because
+/// `Future` uses an associated type, not a type parameter. This trait bridges the gap
+/// by wrapping `Future::poll` behind a type-parameterized interface.
+pub(crate) trait DequeFuture<T> {
```
Comment (sandersaares) [inline packages/futurism/src/deque_future.rs]: DequeFuture and FutureDeque are too confusingly similar. Just rename this to something generalized but distinct like AnonymousFuture or ErasedFuture or UserFuture or similar.

File: packages/futurism/src/future_deque.rs [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+use futures::Stream;
+use infinity_pool::{BlindPool, BlindPooledMut};
+
+use crate::{
+    deque_future::{DequeFuture, PooledCastDequeFuture as _},
+    future_deque_core::{FutureDequeCore, FutureHandle},
+};
+
+thread_local! {
+    static FUTURES_POOL: BlindPool = BlindPool::new();
```
[the remark this answers is not in the document]
Comment (sandersaares) [reply packages/futurism/src/future_deque.rs]: Added inline comments explaining the role, lifecycle, and design rationale for FUTURES_POOL, LOCAL_FUTURES_POOL, and WAKER_META_POOL.

File: packages/futurism/src/future_deque.rs [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+/// futures, see [`LocalFutureDeque`][crate::LocalFutureDeque].
+///
+/// # Implements `Stream`
+///
+/// The [`Stream`] implementation drives all active futures and yields completed results
+/// from the front of the deque. To retrieve results from the back, use [`pop_back`][Self::pop_back]
+/// after driving the deque (e.g. via a `Stream::poll_next` call or by calling it
+/// within an async context).
+#[non_exhaustive]
+pub struct FutureDeque<T> {
```
[the remark this answers is not in the document]
Comment (sandersaares) [reply packages/futurism/src/future_deque.rs]: Added doctest examples on both types showing push/drive/poll_front/pop patterns.

File: packages/futurism/src/waker_meta.rs [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    shared_parent: Arc<Mutex<Waker>>,
+
+    // Self-referential pool handle for cleanup when refcount reaches zero. Set to Some
+    // immediately after pool insertion; None only during the brief construction window.
+    // Uses UnsafeCell because we need interior mutability during bootstrapping (writing
+    // the handle back through a shared reference obtained from as_ref).
+    self_handle: UnsafeCell<Option<RawPooled<Self>>>,
+
+    // Reference to the creating thread's metadata pool for self-cleanup. Waker drops
+    // can happen on any thread; this Arc keeps the pool alive and accessible.
```
[the remark this answers is not in the document]
Comment (sandersaares) [reply packages/futurism/src/waker_meta.rs]: Clarified: Arc is needed because waker drops can happen on any thread after the origin thread has terminated.

File: packages/future_deque/src/future_deque.rs [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+/// with a task context. This polls all activated futures and transitions completed ones to
+/// ready state. Returns `Poll::Ready(())` when all contained futures have completed (or
+/// the deque is empty), and `Poll::Pending` while any futures remain pending. Pushing new
+/// futures resets readiness back to pending.
+///
+/// The deque also implements [`Future`], allowing `.await` to wait for all contained
+/// futures to complete. The deque may be polled again after returning `Poll::Ready(())`
+/// (e.g. after pushing new futures).
+///
+/// [`poll_front`][Self::poll_front] and [`poll_back`][Self::poll_back] combine polling with
```
[the remark this answers is not in the document]
Comment (sandersaares) [reply packages/future_deque/src/future_deque.rs]: Fixed: type-level docs now describe poll_front/poll_back return semantics (Ready(Some(value)), Ready(None), Pending) inline in the Polling the deque section.
